console/network/src/lib.rs

```diff
@@ -228,35 +233,101 @@ pub trait Network:
 
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `N::CONSENSUS_VERSION`
-    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 8];
+    /// Do not read this directly outside of tests, use `N::CONSENSUS_VERSION_HEIGHTS()` instead.
+    const _CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 8];
+
     ///  A list of (consensus_version, size) pairs indicating the maximum number of validators in a committee.
     //  Note: This value must **not** decrease without considering the impact on serialization.
     //  Decreasing this value will break backwards compatibility of serialization without explicit
     //  declaration of migration based on round number rather than block height.
     //  Increasing this value will require a migration to prevent forking during network upgrades.
     const MAX_CERTIFICATES: [(ConsensusVersion, u16); 4];
 
+    /// Returns the list of consensus versions.
+    #[allow(non_snake_case)]
+    #[cfg(not(any(test, feature = "test", feature = "test_consensus_heights")))]
+    fn CONSENSUS_VERSION_HEIGHTS() -> &'static [(ConsensusVersion, u32); 8] {
+        // Initialize the consensus version heights directly from the constant.
+        CONSENSUS_VERSION_HEIGHTS.get_or_init(|| Self::_CONSENSUS_VERSION_HEIGHTS)
+    }
+    /// Returns the list of test consensus versions.
+    #[allow(non_snake_case)]
+    #[cfg(any(test, feature = "test", feature = "test_consensus_heights"))]
+    fn CONSENSUS_VERSION_HEIGHTS() -> &'static [(ConsensusVersion, u32); NUM_CONSENSUS_VERSIONS] {
+        // NOTE: this closure may panic, as it is only called during startup.
+        CONSENSUS_VERSION_HEIGHTS.get_or_init(|| {
+            // Define consensus version heights used for testing.
+            let mut test_consensus_heights = Self::_CONSENSUS_VERSION_HEIGHTS;
+            // Assert that the genesis height is 0.
+            assert_eq!(test_consensus_heights[0].1, 0, "Genesis height must be 0.");
```

Comment:
Why do these checks need to be done for the test heights when it's not done for the real ones?

Reply:
Can there be a helper function that checks all of these for both cases? It would keep the Network trait a bit cleaner as well.

Additionally, this particular check is not done for the heights passed in via ENV variable.

Reply:
Yeah I went back and forth on this actually, cleaned it up further: 58334b36d let me know if this is what you had in mind.